File: app_manager/appmanager/src/main/java/app/editors/manager/ui/fragments/main/DocsTrashFragment.java
package app.editors.manager.ui.fragments.main;

import android.content.Context;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.List;

import app.editors.manager.R;
import app.editors.manager.managers.providers.CloudFileProvider;
import app.editors.manager.mvp.models.base.Entity;
import app.editors.manager.mvp.models.explorer.Item;
import app.editors.manager.mvp.presenters.main.DocsBasePresenter;
import app.editors.manager.ui.activities.main.MainActivity;
import app.editors.manager.ui.dialogs.ContextBottomDialog;
import lib.toolkit.base.managers.utils.UiUtils;

public class DocsTrashFragment extends DocsCloudFragment implements View.OnClickListener {

    private MainActivity mMainActivity;
    private MenuItem mEmptyTrashItem;

    public static DocsTrashFragment newInstance() {
        return new DocsTrashFragment();
    }

    public static final String ID = CloudFileProvider.Section.Trash.getPath();

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try {
            mMainActivity = (MainActivity) context;
        } catch (ClassCastException e) {
            throw new RuntimeException(DocsTrashFragment.class.getSimpleName() + " - must implement - " +
                    MainActivity.class.getSimpleName());
        }
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        init();
    }

    @Override
    public void onPrepareOptionsMenu(@NonNull Menu menu) {
        showMenu(menu);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater menuInflater) {
        super.onCreateOptionsMenu(menu, menuInflater);
        if (isVisible()) {
            showMenu(menu);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.toolbar_item_empty_trash) {
            showQuestionDialog(getString(R.string.trash_dialog_empty_title), null,
                    getString(R.string.dialogs_question_accept_yes), getString(R.string.dialogs_question_accept_no),
                    DocsBasePresenter.TAG_DIALOG_BATCH_EMPTY);
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected boolean onSwipeRefresh() {
        if (!super.onSwipeRefresh()) {
            mCloudPresenter.getItemsById(ID);
            return true;
        }

        return false;
    }

    @Override
    public void onScrollPage() {
        super.onScrollPage();
        mCloudPresenter.setTrashMode(true);
        mCloudPresenter.getItemsById(ID);
        initViews();
    }

    @Override
    public void onClick(View v) {
        mCloudPresenter.moveContext();
    }

    @Override
    public void onItemContextClick(View view, int position) {
        final Item item = (Item) mExplorerAdapter.getItem(position);
        mCloudPresenter.onContextClick(item, position, true);
    }

    @Override
    public void onStateEmptyBackStack() {
        super.onStateEmptyBackStack();
        if (mSwipeRefresh != null) {
            mSwipeRefresh.setRefreshing(true);
        }
        mCloudPresenter.getItemsById(ID);
    }

    @Override
    protected void setMenuMainEnabled(boolean isEnabled) {
        super.setMenuMainEnabled(isEnabled);
        if (mEmptyTrashItem != null) {
            mEmptyTrashItem.setVisible(isEnabled);
        }
        if (mSearchItem != null) {
            mSearchItem.setVisible(isEnabled);
        }
    }

    @Override
    protected boolean isActivePage() {
        return true;
    }

    private void init() {
        mMainActivity.setNavigationButton(true);
    }

    private void showMenu(Menu menu) {
        if (mCloudPresenter.isSelectionMode()) {
            mDeleteItem.setVisible(true);
            mMoveItem.setVisible(true);
            mCopyItem.setVisible(false);
        } else {
            mEmptyTrashItem = menu.findItem(R.id.toolbar_item_empty_trash);
            mEmptyTrashItem.setVisible(false);
            setActionBarTitle("");
            UiUtils.setMenuItemTint(requireContext(), mEmptyTrashItem, R.color.colorWhite);
        }
    }

    private void initViews() {
        if (mRecyclerView != null) {
            mRecyclerView.setPadding(mRecyclerView.getPaddingLeft(), mRecyclerView.getPaddingTop(),
                    mRecyclerView.getPaddingLeft(), 0);
            mCloudPresenter.setTrashMode(true);
            mCloudPresenter.checkBackStack();
        }
    }

    @Override
    public void onContextButtonClick(ContextBottomDialog.Buttons buttons) {
        switch (buttons) {
            case DELETE:
                showQuestionDialog(getString(R.string.trash_popup_delete), mCloudPresenter.getItemTitle(),
                        getString(R.string.dialogs_question_accept_delete), getString(R.string.dialogs_common_cancel_button),
                        DocsBasePresenter.TAG_DIALOG_BATCH_DELETE_CONTEXT);
                break;
            case MOVE:
                mCloudPresenter.moveContext();
                break;
        }
        mContextBottomDialog.dismiss();
    }

    @Override
    public void onDeleteBatch(List<Entity> list) {
        if (list.isEmpty()) {
            setMenuMainEnabled(false);
        }
        super.onDeleteBatch(list);
    }
}

File: app_manager/appmanager/src/main/java/app/editors/manager/ui/fragments/main/DocsTrashFragment.kt
package app.editors.manager.ui.fragments.main

import android.os.Bundle
import android.view.View
import app.documents.core.network.common.contracts.ApiContract
import app.documents.core.network.manager.models.base.Entity
import app.editors.manager.R
import app.editors.manager.managers.tools.ActionMenuItem
import app.editors.manager.mvp.models.filter.RoomFilterType
import app.editors.manager.mvp.models.states.OperationsState
import app.editors.manager.mvp.presenters.main.DocsBasePresenter
import app.editors.manager.ui.dialogs.explorer.ExplorerContextItem
import app.editors.manager.ui.views.custom.PlaceholderViews
import lib.toolkit.base.managers.utils.UiUtils
import lib.toolkit.base.ui.dialogs.common.CommonDialog

class DocsTrashFragment : DocsCloudFragment() {

    private val isArchive: Boolean get() = section == ApiContract.SectionType.CLOUD_ARCHIVE_ROOM

    private val mainPagerFragment: IMainPagerFragment? by lazy {
        requireActivity().supportFragmentManager
            .fragments
            .filterIsInstance<IMainPagerFragment>()
            .firstOrNull()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initViews()
    }

    override fun setMenuMainEnabled(isEnabled: Boolean) {
        super.setMenuMainEnabled(isEnabled)
        searchItem?.isVisible = isEnabled
    }

    override fun onStateMenuSelection() {
        super.onStateMenuSelection()
        deleteItem?.isVisible = !isArchive
    }

    private fun initViews() {
        recyclerView?.let {
            it.setPadding(
                it.paddingLeft, it.paddingTop,
                it.paddingLeft, 0
            )
        }
        explorerAdapter?.isTrash = !isArchive
    }

    override fun onContextButtonClick(contextItem: ExplorerContextItem) {
        when (contextItem) {
            is ExplorerContextItem.Delete -> showDeleteDialog(tag = DocsBasePresenter.TAG_DIALOG_BATCH_DELETE_CONTEXT)
            is ExplorerContextItem.Restore -> {
                if (isArchive) {
                    showRestoreDialog()
                } else {
                    presenter.moveCopyOperation(OperationsState.OperationType.RESTORE)
                }
            }

            is ExplorerContextItem.RoomInfo -> showRoomInfoFragment()
            else -> super.onContextButtonClick(contextItem)
        }
    }

    override fun onDeleteBatch(list: List<Entity>) {
        if (list.isEmpty()) {
            setMenuMainEnabled(false)
        }
        super.onDeleteBatch(list)
    }

    override fun onDeleteMessage(count: Int) {
        onSnackBar(resources.getQuantityString(R.plurals.operation_delete_irretrievably, count))
    }

    override fun showDeleteDialog(count: Int, toTrash: Boolean, tag: String) {
        UiUtils.showQuestionDialog(
            context = requireContext(),
            title = if (count > 0) {
                resources.getQuantityString(R.plurals.dialogs_question_delete_title, count, count)
            } else {
                getString(R.string.dialogs_question_delete_all_title)
            },
            description = resources.getQuantityString(
                R.plurals.dialogs_question_message_delete,
                count
            ),
            acceptTitle = getString(R.string.dialogs_question_accept_delete),
            cancelTitle = getString(R.string.dialogs_common_cancel_button),
            acceptErrorTint = true,
            acceptListener = {
                if (isResumed && isArchive) {
                    presenter.deleteRoom()
                } else if (isResumed && section == ApiContract.SectionType.CLOUD_TRASH) {
                    super.onAcceptClick(CommonDialog.Dialogs.QUESTION, null, tag)
                }
            })
    }

    override fun onResume() {
        super.onResume()
        presenter.isTrashMode = true
    }

    override fun onPause() {
        super.onPause()
        presenter.isTrashMode = false
        mainPagerFragment?.setToolbarInfo(null)
    }

    override fun setToolbarState(state: ToolbarState) {
        if (!isActivePage) return
        if (state is ToolbarState.Trash) {
            mainPagerFragment?.setToolbarInfo(getString(R.string.trash_toolbar_info))
        } else {
            mainPagerFragment?.setToolbarInfo(null)
        }
    }

    override fun getFilters(): Boolean {
        return if (isArchive) {
            val filter = presenter.preferenceTool.filter
            filter.roomType != RoomFilterType.None || filter.author.id.isNotEmpty()
        } else super.getFilters()
    }

    override val actionMenuClickListener: (ActionMenuItem) -> Unit = { item ->
        when (item) {
            ActionMenuItem.Restore -> {
                if (isArchive) {
                    showRestoreDialog()
                } else {
                    presenter.moveCopySelected(OperationsState.OperationType.RESTORE)
                }
            }

            else -> super.actionMenuClickListener(item)
        }
    }

    override fun onPlaceholder(type: PlaceholderViews.Type) {
        if (type == PlaceholderViews.Type.EMPTY && isActivePage) {
            mainPagerFragment?.setToolbarInfo(null)
        }
        val placeholder = when {
            type != PlaceholderViews.Type.EMPTY || !isRoot -> type
            isArchive && presenter.isRegularUser -> PlaceholderViews.Type.EMPTY_ARCHIVE_VIEWER
            isArchive -> PlaceholderViews.Type.EMPTY_ARCHIVE
            else -> PlaceholderViews.Type.EMPTY_TRASH
        }
        super.onPlaceholder(placeholder)
    }

    private fun showRestoreDialog() {
        val count = presenter.getSelectedItemsCount() or 1
        UiUtils.showQuestionDialog(
            context = requireContext(),
            title = resources.getQuantityString(R.plurals.rooms_restore_title, count),
            description = resources.getQuantityString(R.plurals.rooms_restore_desc, count),
            acceptTitle = getString(R.string.trash_snackbar_move_button),
            acceptListener = { presenter.archiveRooms(false) }
        )
    }

    companion object {

        fun newInstance(section: Int, rootPath: String): DocsCloudFragment {
            return DocsTrashFragment().apply {
                arguments = Bundle(2).apply {
                    putString(KEY_PATH, rootPath)
                    putInt(KEY_SECTION, section)
                }
            }
        }
    }
}